cudax/include/cuda/experimental/__memory_resource/memory_pool_base.cuh

```diff
@@ -0,0 +1,399 @@
+  //! true for `cudaMemHandleTypeNone`.
+  //! @param __device_id The id of the device to check for support.
+  //! @param __handle_type An IPC export handle type to check for support.
+  //! @return true if the handle type is supported by cudaDevAttrMemoryPoolSupportedHandleTypes.
+  static void __cuda_supports_export_handle_type(const int __device_id, cudaMemAllocationHandleType __handle_type)
+  {
+    int __supported_handles = static_cast<int>(cudaMemAllocationHandleType::cudaMemHandleTypeNone);
+    if (__handle_type != cudaMemAllocationHandleType::cudaMemHandleTypeNone)
+    {
+      const ::cudaError_t __status =
+        ::cudaDeviceGetAttribute(&__supported_handles, ::cudaDevAttrMemoryPoolSupportedHandleTypes, __device_id);
+      // export handle is not supported at all
+      switch (__status)
+      {
+        case ::cudaSuccess:
+          break;
+        case ::cudaErrorInvalidValue:
+          ::cudaGetLastError(); // Clear CUDA error state
+          ::cuda::__throw_cuda_error(
+            ::cudaErrorNotSupported, "Requested IPC memory handle type not supported on given device");
+        default:
+          ::cudaGetLastError(); // Clear CUDA error state
+          ::cuda::__throw_cuda_error(__status, "Failed to call cudaDeviceGetAttribute");
+      }
+    }
+    if ((static_cast<int>(__handle_type) & __supported_handles) != static_cast<int>(__handle_type))
+    {
+      ::cuda::__throw_cuda_error(
+        ::cudaErrorNotSupported, "Requested IPC memory handle type not supported on given device");
+    }
+  }
+
+  //! @brief  Creates the CUDA memory pool from the passed in arguments.
+  //! @throws cuda_error If the creation of the CUDA memory pool failed.
+  //! @returns The created CUDA memory pool.
+  _CCCL_NODISCARD static cudaMemPool_t
+  __create_cuda_mempool(__memory_location_type __kind, memory_pool_properties __properties, int __id) noexcept
+  {
+    if (__kind == __memory_location_type::__device)
+    {
+      ::cuda::experimental::__verify_device_supports_stream_ordered_allocations(__id);
+      __memory_pool_base::__cuda_supports_export_handle_type(__id, __properties.allocation_handle_type);
+    }
+
+    ::cudaMemPoolProps __pool_properties{};
+    __pool_properties.allocType   = ::cudaMemAllocationTypePinned;
+    __pool_properties.handleTypes = ::cudaMemAllocationHandleType(__properties.allocation_handle_type);
+    if (__kind == __memory_location_type::__device)
+    {
+      __pool_properties.location.type = ::cudaMemLocationTypeDevice;
+      __pool_properties.location.id   = __id;
+    }
+    else
+    {
+#if _CCCL_CUDACC_AT_LEAST(12, 6)
+      // Construct on NUMA node 0 only for now
+      __pool_properties.location.type = ::cudaMemLocationTypeHostNuma;
+      __pool_properties.location.id   = __id;
+#else // _CCCL_CUDACC_BELOW(12, 6)
+      _CCCL_ASSERT(false, "Host pinned memory pools unavailable");
+      _CUDA_VSTD_NOVERSION::terminate();
+#endif // _CCCL_CUDACC_AT_LEAST(12, 6)
+    }
```

Comment:
Nitpick: I would prefer if this would be a switch statement with `__memory_location_type::__device` as the default